InvocationHandlerDelegation false positive?

## What happened?

With baseline 3.23.0, the following code trips over

```
protected Object handleInvocation(Object _proxy, Method method, Object[] args) throws Throwable {
            try {
                Object ret = method.invoke(delegate, args);
                log(AuditResult.SUCCESS);
                return ret;
            } catch (InvocationTargetException | UndeclaredThrowableException e) {
                log(AuditUtil.getResultForThrowable(e.getCause()));
                throw e.getCause();
            } catch (Throwable e) {
                log(AuditUtil.getResultForThrowable(e));
                throw e;
            } finally {
                context.clear();
            }
        }
```

Error produced:

```
> Task :audit-api:compileJava
/Volumes/git/foundry-audit/audit-api/src/main/java/com/palantir/foundry/audit/AuditLoggingProxy.java:41: warning: [InvocationHandlerDelegation] InvocationHandlers which delegate to another object must catch and unwrap InvocationTargetException, otherwise an UndeclaredThrowableException will be thrown each time the delegate throws an exception.
                Object ret = method.invoke(delegate, args);
                                          ^
  This check is intended to be advisory. It's fine to @SuppressWarnings("InvocationHandlerDelegation") in certain cases, but is usually not recommended.
    (see https://github.com/palantir/gradle-baseline#baseline-error-prone-checks)
```

## What did you want to happen?


I _think_ this code is fine and it's just a straight-up false positive?? (possibly multi-branch catch block?)

--- a/baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/InvocationHandlerDelegation.java
+++ b/baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/InvocationHandlerDelegation.java
@@ -26,6 +26,10 @@ import com.google.errorprone.matchers.Description;
 import com.google.errorprone.matchers.Matcher;
 import com.google.errorprone.matchers.Matchers;
 import com.google.errorprone.matchers.method.MethodMatchers;
+import com.google.errorprone.predicates.TypePredicate;
+import com.google.errorprone.predicates.TypePredicates;
+import com.google.errorprone.predicates.type.DescendantOf;
+import com.google.errorprone.suppliers.Suppliers;
 import com.google.errorprone.util.ASTHelpers;
 import com.sun.source.tree.ConditionalExpressionTree;
 import com.sun.source.tree.ExpressionTree;
@@ -37,6 +41,7 @@ import com.sun.source.tree.MethodTree;
 import com.sun.source.tree.NewClassTree;
 import com.sun.source.tree.Tree;
 import com.sun.source.tree.TryTree;
+import com.sun.tools.javac.code.Type;
 import java.lang.reflect.InvocationHandler;
 import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
@@ -83,8 +88,22 @@ public final class InvocationHandlerDelegation extends BugChecker implements Bug
     private static final Matcher<Tree> CONTAINS_UNWRAP_THROWABLE =
             Matchers.contains(ExpressionTree.class, UNWRAP_THROWABLE);
 
+    private static TypePredicate IS_ITE_SUBTYPE =
+            new DescendantOf(Suppliers.typeFromClass(InvocationTargetException.class));
+
+    private static TypePredicate IS_ITE_UNION = (TypePredicate) (type, state) -> {
+        if (type.isUnion()) {
+            for (Type unionType : MoreASTHelpers.expandUnion(type)) {
+                if (IS_ITE_SUBTYPE.apply(unionType, state)) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    };
+
     private static final Matcher<ExpressionTree> UNWRAP_ITE = MethodMatchers.instanceMethod()
-            .onDescendantOf(InvocationTargetException.class.getName())
+            .onClass(TypePredicates.anyOf(IS_ITE_SUBTYPE, IS_ITE_UNION))
             // getTargetException is deprecated, but does work correctly.
             .namedAnyOf("getCause", "getTargetException")
             .withParameters();

--- a/baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/InvocationHandlerDelegation.java
+++ b/baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/InvocationHandlerDelegation.java
@@ -88,10 +88,10 @@ public final class InvocationHandlerDelegation extends BugChecker implements Bug
     private static final Matcher<Tree> CONTAINS_UNWRAP_THROWABLE =
             Matchers.contains(ExpressionTree.class, UNWRAP_THROWABLE);
 
-    private static TypePredicate IS_ITE_SUBTYPE =
+    private static final TypePredicate IS_ITE_SUBTYPE =
             new DescendantOf(Suppliers.typeFromClass(InvocationTargetException.class));
 
-    private static TypePredicate IS_ITE_UNION = (TypePredicate) (type, state) -> {
+    private static final TypePredicate IS_ITE_UNION = (TypePredicate) (type, state) -> {
         if (type.isUnion()) {
             for (Type unionType : MoreASTHelpers.expandUnion(type)) {
                 if (IS_ITE_SUBTYPE.apply(unionType, state)) {

--- a/baseline-error-prone/src/test/java/com/palantir/baseline/errorprone/InvocationHandlerDelegationTest.java
+++ b/baseline-error-prone/src/test/java/com/palantir/baseline/errorprone/InvocationHandlerDelegationTest.java
@@ -134,6 +134,27 @@ class InvocationHandlerDelegationTest {
     }
 
     @Test
+    void testInvocationHandler_catchUnion() {
+        helper().addSourceLines(
+                        "Test.java",
+                        "import java.lang.reflect.InvocationHandler;",
+                        "import java.lang.reflect.Method;",
+                        "import java.lang.reflect.InvocationTargetException;",
+                        "import java.lang.reflect.UndeclaredThrowableException;",
+                        "final class Test implements InvocationHandler {",
+                        "  @Override",
+                        "  public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {",
+                        "    try {",
+                        "      return method.invoke(this, args);",
+                        "    } catch (InvocationTargetException | UndeclaredThrowableException e) {",
+                        "      throw e.getCause();",
+                        "    }",
+                        "  }",
+                        "}")
+                .doTest();
+    }
+
+    @Test
     void testCorrectInvocationHandler_lambda() {
         helper().addSourceLines(
                         "Test.java",

--- a/changelog/@unreleased/pr-1499.v2.yml
+++ b/changelog/@unreleased/pr-1499.v2.yml
@@ -1,0 +1,5 @@
+type: fix
+fix:
+  description: 'fix #1498: Fix InvocationHandlerDelegation false positive'
+  links:
+  - https://github.com/palantir/gradle-baseline/pull/1499

--- a/changelog/@unreleased/pr-1499.v2.yml
+++ b/changelog/@unreleased/pr-1499.v2.yml
@@ -1,5 +1,5 @@
 type: fix
 fix:
-  description: 'fix #1498: Fix InvocationHandlerDelegation false positive'
+  description: Fix InvocationHandlerDelegation false positive
   links:
   - https://github.com/palantir/gradle-baseline/pull/1499

